File: server/tools/check_environment.js
// The pull path into Environment awareness (root CLAUDE.md's Operational
// Awareness item 5) — a live, real picture of the machine Jarvis runs on
// and what it's currently able to reach, so a decision about whether to
// START something can be made BEFORE committing to it rather than finding
// out mid-task. Distinct from check_myself.js (that's about whether Jarvis
// ITSELF is malfunctioning; this is about the surrounding system).
//
// core:true (no reliable search-intent text — "can you handle a heavy task
// right now" doesn't map to a capability search) + meta:true
// (live-conversation-only; a background Job's own turn making this decision
// for itself would need a different, narrower check, not this one).

import { sampleNow } from '../ops/environment/sampler.js';
import { fullReachability } from '../ops/environment/reachability.js';
import { checkForAnomaly } from '../ops/environment/baseline.js';

export default {
  name: 'check_environment',
  core: true,
  meta: true,
  description:
    'Check the real, live state of the machine and services you depend on before starting something that needs them — CPU/memory ' +
    'load right now, whether a specific model/connector/voice service is actually reachable, and whether load has been unusually ' +
    'high or climbing lately. Use this before committing to a heavy or long-running task, not just when asked directly about system health.',
  parameters: {
    type: 'object',
    properties: {},
    required: [],
  },
  async run() {
    const load = sampleNow({ record: false });
    const reachability = fullReachability();
    const anomaly = checkForAnomaly();

    return {
      ok: true,
      loadRightNow: {
        cpuPct: load.cpuPct,
        memFreePct: load.memFreePct !== null ? Math.round(load.memFreePct * 10) / 10 : null,
        jarvisOwnMemoryBytes: load.rssBytes,
      },
      unusualLoad: anomaly ? { summary: anomaly.summary } : null,
      reachability,
    };
  },
};

File: backend/jarvis/tools/check_environment.py
"""Can this actually be done right now, on this machine, at this moment?

A different question from "is Jarvis broken" (`check_my_health`) and from "am I
any good at this" (`check_myself`) — those are about Jarvis; this is about the
conditions it is working in. Answering "sure, I'll transcode that" while the
machine has been pinned at 98% for ten minutes is the failure this exists to
prevent.
"""

from __future__ import annotations

from typing import Any

from ..capabilities import CapabilitySpec, Risk


def _run() -> dict[str, Any]:
    from ..ops.environment import baseline, reachability, sampler

    reading = sampler.read_now()
    anomaly = baseline.check_for_anomaly()
    reach = reachability.snapshot()

    load: dict[str, Any] = {
        "cpuPct": None if reading["cpuPct"] is None else round(reading["cpuPct"], 1),
        "memoryFreePct": None if reading["memFreePct"] is None else round(reading["memFreePct"], 1),
        "cpuCount": reading["cpuCount"],
    }
    if reading["cpuPct"] is None:
        # A percentage is a measurement over an interval, and the first reading
        # of a process has no interval yet. Unknown, not idle.
        load["note"] = "CPU use is not known yet — this is the first reading since starting."

    return {
        "ok": True,
        "load": load,
        "unusual": anomaly["summary"] if anomaly else None,
        "models": reach["models"],
        "voice": reach["voice"],
        "note": ("Real readings from this machine, plus what is actually reachable. "
                 "If nothing is unusual, say so plainly rather than hedging."),
    }


SPEC = CapabilitySpec(
    id="builtin.check_environment",
    name="check_environment",
    description=("Check what this computer and Jarvis's own connections can handle right now — "
                 "how loaded the machine is, whether load has been unusually high, and which "
                 "models and voices are actually reachable. Use it before taking on something "
                 "heavy, or when asked whether you can do something right now."),
    input_schema={"type": "object", "properties": {}, "required": []},
    risk=Risk.LOW,
    handler=_run,
    timeout_s=15.0,
    tags=frozenset({"core", "meta"}),
)
